### PR TITLE
Optimizing Student Progress Report page for screen readers

### PR DESCRIPTION
**Pull request by radinamatic**
#### UPDATE August 5 

This is technically finished, and I used `generaterealdata` function to test it with the learners from Wilson Elementary... :)
I can say I feel satisified, so, **review and merge away**! :)

##### UPDATE August 4
This is technically finished, but I'd like to test it more thoroughly with more content (waiting for the #4135 & #4157 ). So, **don't merge yet**! :)
  
  

Still **WIP**, but it's a lot of fun though, like trying to fit puzzle pieces... :)

##### playlist-progress-container.handlebars  
- [x] Videos
- [x] Exercises
- [x] Quizzes (We don't have those yet, right?)  -  *Oh Quizzes, we hardly knew ye!* by @rtibbles 
- [x] `sr-only` field for 0% progress bar for videos 
- [x] `sr-only` field for 0% progress bar for exercises 
- [x] Conditional logic for screen reader output in case of topics with only videos (& no exercises) or only exercises (& no videos).

- [x] I felt that the output `You are still working on 100%` was not sufficiently clear, so I added another condition to transform those cases in `You are still working on all of them` for both videos and exercises.

##### playlist-progress-details.handlebars
- [x] To avoid headaches about the support for `popover` class being flaky among different screen readers, I opted for rendering that whole section invisible with `aria-hidden="true"`, and then adding a new `sr-only` section with the same info for non-sighted users.

### CODE REVIEW

**Comment by aronasorman** [comment]
Merge conflict :)

**Comment by radinamatic** [comment]
Is the conflict due to **browserify** that got merged yesterday?

**Comment by rtibbles** [comment]
Almost certainly. The files got moved, and I thought git had registered that properly, but perhaps not? :(

**Comment by radinamatic** [comment]
Perfect moment to learn more about solving git conflicts, it seems... :P

**Comment by radinamatic** [comment]
I'm using this logic for producing `sr-only` output in cases where exercises are in progress (I tried with `{{#if ex_status.inprogress}}`, but that condition is not producing the `div` with the progress bar):

```
{{#if ex_pct_incomplete}}
    <div class="progress-bar inprogress" title="% In-Progress" style="width: {{ex_pct_incomplete}}%">
        <span class="sr-only">{{_ "You are still working on " }}</span>
        <span>{{ex_pct_incomplete}}%</span>
    </div>
{{/if}}
```
However, I have a case of a topic where there is only 1 exercise, which produces the 100% progress bar, but gives the screen reader output `You are still working on 100%`, which should indicate that it is in `inprogress` category.

angular

Logically, `sr-only` output makes sense, as the student is still working on the only one exercise in the topic, but visually, the 100%  **green** progress bar gives the impression that the exercise is complete, and they can only realize it's not if they open the progress detail below. 

Maybe for incomplete exercises the progress bar could be colored orange or yellow?

**Comment by radinamatic** [comment]
**UPDATE** August 4: This is technically finished, but I'd like to test it more thoroughly with more content (waiting for the #4135 & #4157 ).

**Comment by radinamatic** [comment]
Almost there... :)

progress-report-sr2  

NVDA output:
```
Heading level two clickable link 
What is addition? What is subtraction? 
In this topic there are two videos. You haven't started to watch videos! 
There are two exercises. You are still working on all of them. 

Heading level two clickable link 
Converting between fractions and decimals 
In this topic there are ten videos. Of those you have completed forty percent. You are still working on sixty percent.
There are three exercises. You are still working on all of them.
```

preogress-report-sr

NVDA output:
```
Heading level two clickable link 
Conceptualizing decimals and place notation 
In this topic there are seven videos. Of those you have completed fifty-seven percent. You are still working on forty-two percent. 
There are three exercises. You are still working on all of them.

Heading level three clickable link 
Decimal place value 
Status In progress. For this video you completed forty-two percent

Heading level three clickable link 
Decimals writing it out in words 
Status Completed. For this video you completed one hundred percent

Heading level three clickable link 
Decimals writing it out in words two 
Status In progress. For this video you completed eighty percent

Heading level three clickable link 
Decimals expanding out place value 
Status Completed. For this video you completed one hundred percent 

Heading level three clickable link 
Decimals writing it out as a number 
Status Completed. For this video you completed one hundred percent

Heading level three clickable link 
Writing and interpreting decimals 
Status In progress. For this exercise your streak is ten percent
...
```

File: kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars [hunk 1]
```diff
@@ -1,28 +1,44 @@
...
                     <div class="progress-bar inprogress" title="% In-Progress" style="width: {{vid_pct_started}}%">
+                    {{#ifcond vid_pct_started "<" 100}}
+                        <span class="sr-only">{{_ "You are still working on " }}</span>
                         <span>{{vid_pct_started}}%</span>
+                    {{/ifcond}}
+                    {{#ifcond vid_pct_started "===" 100}}
+                        <span class="sr-only">{{_ "You are still working on all of them." }}</span>
+                        <span aria-hidden="true" role="presentation">{{vid_pct_started}}%</span>
+                    {{/ifcond}}
                     </div>
```
Comment (radinamatic) [inline kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars]: I admit the condition above might be nitpicking, but I do find the `You are still working on 100%` insufficiently clear as a message for the user.

**Comment by radinamatic** [comment]
I also have a suggestion for background color for `inprogress` state, maybe a yellowish one (#e2e880) would render the "unfinished" message better than the previous light-green:

@jtamiace @rtibbles 

in-progress

**Comment by rtibbles** [comment]
Will be looking at this by EOD Friday 13th August.

**Comment by radinamatic** [comment]
:+1:

File: kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars (hunk 1 above)
Comment (rtibbles) [inline kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars]: No, makes sense. I feel like reviewing this code would be easier with a screen reader!

File: kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars [hunk 2]
```diff
@@ -60,13 +92,17 @@
     <div class="row progress-block">
         <div class="col-xs-2">
             <div class="progress-indicator progress-indicator-lg center-block {{quiz_status}}">
+                <span class="sr-only">{{_ "There are " }}</span>
                 <span class="sidebar-icon icon-Quiz icon-label icon-label-lg"></span>
+                <span class="sr-only">{{_ " quizes." }}</span>
```
Comment (rtibbles) [inline kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars]: typo - 'quizzes'

**Comment by rtibbles** [comment]
Looks excellent. One tiny nit pick in spelling, then this is good to merge.

File: kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars (hunk 1 above)
Comment (radinamatic) [inline kalite/coachreports/static/js/coachreports/student_progress/hbtemplates/playlist-progress-container.handlebars]: That was the whole point! :)

**Pull request merged**